content/vendor/guides/persistent-datastores.md

```diff
@@ -0,0 +1,940 @@
+      - env:
+        - name: PGDATA
+          value: /var/lib/postgresql/data/pgdata
+        # create a db called "postgres"
+        - name: POSTGRES_DB
+          value: postgres
+        # create admin user with name "postgres"
+        - name: POSTGRES_USER
+          value: postgres
+        # use admin password from secret
+        - name: POSTGRES_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              key: DB_PASSWORD
+              name: postgres-secret
+        image: postgres:10
+        name: postgres
+        volumeMounts:
+        - mountPath: /var/lib/postgresql/data
+          name: pgdata
+      volumes:
+      - name: pgdata
+        persistentVolumeClaim:
+          claimName: pgdata
+  volumeClaimTemplates:
+  - metadata:
+      name: pgdata
+    spec:
+      accessModes:
+      - ReadWriteOnce
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+Finally, lets add a Service object so we can route traffic to our postgres instance, again using `kots.io/when` to conditionally include this resource.
+
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: postgres
+  labels:
+    app: pg-provider
+  annotations:
+    kots.io/when: '{{repl ConfigOptionEquals "postgres_type" "embedded_postgres" }}'
+spec:
+  ports:
+  - port: 5432
+  selector:
+    app: pg-provider
+  type: ClusterIP
+```
+
+### Validating the embedded Database
+
+Once you've added the these resources, you can push a new release and update in kotsadm. You should see the following in the deployment logs:
+
+![Embedded PG Deployed](/images/guides/kots/embedded-pg-deployed.png)
+
+We should now see an instance of postgres running in our namespace as well. The consumer may still be crashlooping, but we can see the error is different now:
+
+```text
+$ kubectl logs -l app=pg-consumer 
+psql: FATAL:  password authentication failed for user "postgres"
+```
+
+This is because we still need to deliver the generated password to our workload pod. In `pg-consumer.yaml`, we'll remove this section
+
+```yaml
+            - name: DB_PASSWORD
+              value: postgres
+```
+
+and replace it with
+
+```yaml
+            - name: DB_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: postgres
+                  key: DB_PASSWORD
+```
+
+The full Deployment should now look like:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: pg-consumer
+spec:
+  selector:
+    matchLabels:
+      app: pg-consumer
+  template:
+    metadata:
+      labels:
+        app: pg-consumer
+    spec:
+      containers:
+        - name: pg-consumer
+          image: 'postgres:10'
+          # connect to the database every 20 seconds
+          command:
+            - /bin/sh
+            - -ec
+            - |
+              while :; do 
+                 sleep 20
+                 PGPASSWORD=${DB_PASSWORD} \
+                 psql --host ${DB_HOST} \
+                      --port ${DB_PORT} \ 
```

Comment:
Found it! There is a trailing whitespace on the end of this line. If someone copies this in, it will break the pg-consumer pod with an error saying 'root user not found' or something.

Reply:
oh wow nice thanks.